### PR TITLE
Fix for crash on Linux on sources with surrogate characters

### PR DESCRIPTION
**Pull request by notorca**
Because of the bug https://bugs.swift.org/browse/SR-5971 SwiftLint
crashes when linting on Linux files like https://github.com/xmartlabs/Eureka/blob/master/Example/Example/ViewController.swift

### CODE REVIEW

**Comment by notorca** [comment]
This is temporary solution, until Foundation for linux is not fixed.

**Comment by marcelofabri** [comment]
@notorca Can you add an unit test that would fail without this patch?

**Comment by notorca** [comment]
@marcelofabri done

**Comment by SDGGiesbrecht** [comment]
The technical terms mentioned here have been confused. There is no such thing as a “surrogate character”. (Did you mean “surrogate code point”? But if there is a surrogate code point left after string initialization, then the source code was malformed Unicode to begin with.)

SR‐3215, SR‐3282 report unsolved bugs in that initializer involving only ASCII. On the other hand, I use `CharacterSet` with Unicode often on Linux and have not noticed any issues.

I would like to investigate further to make sure the problem is correctly understood. I will report back.

**Comment by notorca** [comment]
@SDGGiesbrecht you can try this on linux:
```
import Foundation
let cset = CharacterSet(charactersIn:"\u{10000}") 
```

**Comment by notorca** [comment]
Here is the fix for Foundation https://github.com/apple/swift-corelibs-foundation/pull/1471

**Comment by SDGGiesbrecht** [comment]
@notorca, yes, you are correct.

At first I had read [SR‐5971](https://bugs.swift.org/browse/SR-5971) that you linked to, where the hypothesis involved `sizeof(UniChar)`. That hypothesis was clearly wrong.

Then I looked at [`_CFCharacterSetInitWithCharactersInString`](https://github.com/apple/swift-corelibs-foundation/blame/a4ca69eb37dbff36a1f72a125553f6dc7a4385b4/CoreFoundation/String.subproj/CFCharacterSet.c#L1399-L1436), saw that it forwarded to `CFCharacterSetAddCharactersInString` just like your patch here, and that it has been unchanged for two years (i.e. not fixed since last release).

That was why I was suspicious at first.

But you are correct that Foundation’s current implementation `NULL`ifies the pointer before re‐trying with the other initializer. 😱

Both your patches (this one and the other in Foundation) look good to me. Good catch.

**Comment by SDGGiesbrecht** [comment]
@notorca,

From your change log entry:
> Fix linux crash on sources with surrogate pair emojis as variable names

Actually, you have fixed a lot more than that. Not just in variables, but several different token types, and not just emoji, but over 94% of all Unicode scalars.

If you want, you can change it to something like, “Fixed Linux crashes on source files with Unicode scalars from supplementary planes.”

But you can leave it the way you have it if you would rather.

File: CHANGELOG.md [hunk 1]
```diff
@@ -17,6 +17,9 @@
   [Marcelo Fabri](https://github.com/marcelofabri)
   [#2050](https://github.com/realm/SwiftLint/issues/2050)
 
+* Fix linux crash on sources with surrogate pair emojis as variable names
```
Comment (marcelofabri) [inline CHANGELOG.md]: requires period and two trailing spaces as described in CONTRIBUTING.md: https://github.com/realm/SwiftLint/blob/master/CONTRIBUTING.md#tracking-changes

File: Source/SwiftLintFramework/Extensions/CharacterSet+SwiftLint.swift [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+//  Copyright © 2018 Realm. All rights reserved.
+//
+
+import Foundation
+
+public extension CharacterSet {
+    init(safeCharactersIn string: String) {
+#if os(Linux)
+        self.init()
+        self.insert(charactersIn: string)
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Extensions/CharacterSet+SwiftLint.swift]: unneeded `self`

File: Source/SwiftLintFramework/Extensions/CharacterSet+SwiftLint.swift [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+//  CharacterSet+SwiftLint.swift
+//  SwiftLint
+//
+//  Created by Cyril Lashkevich on 3/11/18.
+//  Copyright © 2018 Realm. All rights reserved.
+//
+
+import Foundation
+
+public extension CharacterSet {
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Extensions/CharacterSet+SwiftLint.swift]: this should be `internal`, not `public`

File: Source/SwiftLintFramework/Extensions/CharacterSet+SwiftLint.swift [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    init(safeCharactersIn string: String) {
+#if os(Linux)
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Extensions/CharacterSet+SwiftLint.swift]: can you please link the SR issue here so we can track when that's fixed?

**Comment by notorca** [comment]
@marcelofabri Done,

**Comment by marcelofabri** [comment]
@notorca Thanks!

**Comment by jpsim** [comment]
Nice sleuthing here folks! One thing someone could do if they’d like is to add a custom rule to lint for our uses of `CharacterSet(charactersIn:)` to ensure we use the safe version.

**Pull request merged**